src/encoder/text.rs

```diff
@@ -32,23 +34,33 @@ impl Encoder for TextEncoder {
             // Fail-fast checks.
             check_metric_family(mf)?;
 
+            // Write `# HELP` header.
             let name = mf.get_name();
             let help = mf.get_help();
             if !help.is_empty() {
-                writeln!(writer, "# HELP {} {}", name, escape_string(help, false))?;
+                writer.write_all(b"# HELP ")?;
+                writer.write_all(name.as_bytes())?;
+                writer.write_all(b" ")?;
+                writer.write_all(escape_string(help, false).as_bytes())?;
+                writer.write_all(b"\n")?;
```

Comment:
This pattern looks like a [vectored write](https://en.wikipedia.org/wiki/Vectored_I/O). Did you consider directly going for https://doc.rust-lang.org/std/io/trait.Write.html#method.write_vectored here?

Reply:
`write_vectored` won't guarantee that everything in the `&[IoSlice]` are written out, unlike `write_all`. In fact, the default implementation of `write_vectored` on non-files will only write out the first non-empty slice. This means these 5 lines will need to become a loop if we use `write_vectored` instead.

Reply:
For `<Vec as Write>` it would have the benefit of reserving enough space for all slices at once: 

```rust
    #[inline]
    fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
        let len = bufs.iter().map(|b| b.len()).sum();
        self.reserve(len);
        for buf in bufs {
            self.extend_from_slice(buf);
        }
        Ok(len)
    }
```

https://doc.rust-lang.org/src/std/io/impls.rs.html#337-364

Not sure how to handle the case where not everything was written at once efficiently yet. Will need to put more thoughts into that. Not sure whether we just want to copy https://github.com/rust-lang/rust/pull/70612/.

Reply:
@kennytm I missed that nuance, thanks for sharing that!

Reply:
@mxinden I'm fine deferring this pending libstd stabilization.